### DISCUSSION TITLE
Track kills per map ?

### DISCUSSION THREAD

**Discussion by GrimDawn-max** [Lailloken/Exile-UI#120] [Ideas & Feedback]
Could you add a feature that tracks kills per map and include it in the loot stats export file. The stat exists manually of course via the /kills command but you also have to remember to execute it before you kill anything in a map and again at the end so an automated way to do so would be handy.

**Comment by Lailloken** [comment]
The output of the /kills command does indeed get logged in the client.txt, I just checked that, but it's against the TOS to trigger a macro based on file changes or screen-reading. The best I could do is make it an optional feature that adds an extra button to the panel with the map-name and timer. I could highlight it somehow whenever the user enters a new instance to signal that it needs to be clicked to trigger the /kills macro and log the starting kill-count. Then, each time the user leaves the map device the button would be highlighted again to signal that it needs to be clicked to trigger the macro again to log the kill-count at the end. (EDIT: of course, this would only have to be done once, whenever you know the map is actually completed, not everytime you leave for something else). The difference between the values would represent the kill-count for that specific map which would then be logged alongside the other data. The highlighting on the button would have to be noticeable enough to not forget/overlook it. Sounds good in theory, let me know if you can think of improvements.

**Comment by Lailloken** [comment]
Test-version is up if you want to do some testing: https://github.com/Lailloken/Lailloken-UI/archive/refs/heads/dev.zip

Again, back up your folder before applying this update. There's a new checkbox in the settings menu, and there are also instructions on how the kill tracker works. I was only able to do limited testing, so it might be a bit rough here and there.

**Reply by GrimDawn-max** [reply]
I tested the feature. I'm not sure I'm testing it correctly as I'm not getting a kill count in the export file (the field is there but it is blank) and the timer button flashes green the entire map.

I do the following
- open map, enter map, left click timer button once, execute /kills command. The timer button continues to flash green the entire map
- clear map, execute /kills command again, return to HO, timer stops with no flashing and is green
- perform loot tracker, export stats

All data is there except content and kills fields. Both blank. The /kills command output is in the Client.txt file.
If I'm doing things correctly then there's a bug. If I'm not doing things correctly you may need to provide a short video to show what needs to be done.

**Reply by Lailloken** [reply]
https://www.youtube.com/watch?v=MhkOm5P_ERg (still being processed)
Clicking the timer should automatically send the /kills command. The panel staying green at the end is a bug.

**Comment by GrimDawn-max** [comment]
Sure. I can test it later today and let you know if there are any issues. Thanks for putting it together !

**Comment by GrimDawn-max** [comment]
I’ve tried right click, left click and middle button. None of them work for me. I don’t get the activated message or the kills output on screen.

I have kill tracker enabled in the config menu

**Reply by Lailloken** [reply]
Are you still using a VM? Can you clear your clipboard, then enter a new map (so that the panel starts flashing), and click the timer. Upon clicking the timer, the clipboard should contain /kills, and the script then waits for left-click to be released (safety-measure to avoid the poe client losing focus). Maybe it's not reading the state of the left mouse-button correctly.

**Reply by Lailloken** [reply]
I updated the dev build and added a beep when the timer is clicked. This way, we can tell if the timer is not functioning as a button or if whatever is triggered by it is malfunctioning.

**Comment by GrimDawn-max** [comment]
I am using a VM. Don’t have a native Windows box I’m afraid. I have to step out for a bit. Will try again in an hour or so. Do I left-click on the timer window after entering map just the once for the popup that /kills has been executed ?

**Reply by Lailloken** [reply]
Yes, it's just there to have a starting value X. Clicking it again at any point afterwards will set a value Y. Whenever data is saved, Y-X will be logged as the kill-count for the map. I could've made it so Y becomes the new X for the new map automatically, but that would cause problems if the user were to go do side-content in-between maps (delve, lab, etc.). So I opted for setting X manually every map.

I'll be offline from now (sleep), so there won't be any response for a bit. But the situation seems to be some sort of compatibility problem: if the beep works, the code triggered by clicking the timer is bugged. If it doesn't work, the timer doesn't work as a button for some reason. The clip was taken with a clean install of the dev build, so it's strange that it's not working correctly on your end.

**Reply by GrimDawn-max** [reply]
It's working now. I hear the beep and the output of the kills command is displayed on screen. I don't know if clearing the clipboard had any effect. It may have been me not clicking in the correct area of the timer button to begin with. 

What I did notice is that you have to left click on a specific section of the timer button. You have to click on the timer number field. Anything to the left of that will not activate the command. Maybe it's intentionally that way. That's fine. 

The output of the kill command is still not showing in the export file. Everything else is (time, date, map name, portals, deaths and loot). I'll try it without the ctrl+click to include loot and see if the kill appears then.

**Reply by GrimDawn-max** [reply]
Excluding the ctrl+click for inventory didn't make a difference. The kill count still shows as zero. Other than that everything is working as expected.

By the way -  Iassume the kill count in the export file is a total for the map (subtract the last kill command from the first one) not just a dump of the command output from the Client.txt file ?

**Reply by Lailloken** [reply]
> What I did notice is that you have to left click on a specific section of the timer button. You have to click on the timer number field.

Yeah, It's just for the time being because I didn't quite know if I would add a dedicated button or counter to the panel.
<br>

> The output of the kill command is still not showing in the export file

Are you sure you're not mixing up install folders?
![image](https://user-images.githubusercontent.com/61888437/202097693-32ddf97f-69a4-4dae-838d-2b48eda78d53.png)

**Reply by GrimDawn-max** [reply]
Yes. I only have the one folder. Mapping tracker.csv is only in one place. My file looks the same as yours except all the kills totals are 0

<img width="733" alt="MapOutput" src="https://user-images.githubusercontent.com/64788480/202205209-f8310659-faf1-4c56-8762-3784be3d6b20.png">

**Reply by Lailloken** [reply]
OK, there has to be a problem in this chain:
- clicking the timer when the panel is flashing sends the /kills command and sets kills to 0 -> do you get the popup "kill-tracker activated"?

- clicking it again afterwards will send the command again and refresh the kill-count -> do you get the popup "kills: X"?
- after the run is complete and the data logged, preliminary data is stored in ini\map tracker log.ini -> does the data for a run have "kills=X"?
- in the in-game log-viewer the run should be highlighted in cyan and long-clicking it should display additional information about the run -> does it show the kill-count?
- exporting either a specific day or the complete data should contain that specific kill-count -> is that the case?

Please check this list one step at a time in order to find out where the problem is. It's either that the kill-count is not parsed correctly from the client.txt, not saved into the ini, or not read from the ini.

**Reply by GrimDawn-max** [reply]
OK, there has to be a problem in this chain:

 * clicking the timer when the panel is flashing sends the /kills command and sets kills to 0 -> do you get the popup "kill-tracker activated"?
>Yes
* clicking it again afterwards will send the command again and refresh the kill-count -> do you get the popup "kills: X"?
>Yes
* after the run is complete and the data logged, preliminary data is stored in ini\map tracker log.ini -> does the data for a run have "kills=X"?
>Yes but kills=0
* in the in-game log-viewer the run should be highlighted in cyan and long-clicking it should display additional information about the run -> does it show the kill-count?
>Haven't checked yet
* exporting either a specific day or the complete data should contain that specific kill-count -> is that the case?
>Yes but it is zero
 
 Please check this list one step at a time in order to find out where the problem is. It's either that the kill-count is not parsed correctly from the client.txt, not saved into the ini, or not read from the ini.
> It would appear it is not parsing from the Client.txt file to the ini file (since that is zero). The Client.txt files has the kills command entries and they are not zero

**Reply by Lailloken** [reply]
That's so weird. There's no reason for the value in the "kills: X" popup and the one in the ini to be different, they are one and the same. Could you try re-applying the update? Alternatively, extract the dev build into a clean, separate folder and try it from there since the trackers don't need much of a setup.

**Reply by GrimDawn-max** [reply]
> That's so weird. There's no reason for the value in the "kills: X" popup and the one in the ini to be different, they are one and the same. Could you try re-applying the update? Alternatively, extract the dev build into a clean, separate folder and try it from there since the trackers don't need much of a setup.

I downloaded the install from the link in this post, installed it in a completely separate folder and ran the application as usual
It executes the commands as you would expect. 

>in the in-game log-viewer the run should be highlighted in cyan and long-clicking it should display additional information about the run > -> does it show the kill-count?

The colour is cyan but there is no was kill-count in the listing or the ini file

The map tracker ini file has 
track side-areas=1
enable loot tracker=1
enable kill tracker=1

The map tracker log file looks like this
<img width="1124" alt="MapTrackFile" src="https://user-images.githubusercontent.com/64788480/202276502-7ee00823-0276-4244-bb69-e27a36cdb484.png">

The /kills commands output in the chat window looks like this (as it does in the Client.txt file)
<img width="1070" alt="KillOutput" src="https://user-images.githubusercontent.com/64788480/202276727-f88e2442-ebd6-4ce2-bb8f-bee33b7a149d.png">

It does not matter if I trigger the last /kills command as the last thing after completing the map or wait till back in the hideout. The output of kills is still zero

I have AHK scripts do weird things on me before. It can be a bit quirky :-) and a pain to troubleshoot. Based on the chain you outlined it's not reading the Client.txt file for the counts, or it's reading the same value twice, which would result in a zero instead of a null. Can you add debug statements to test if is reading and recording from the Client.txt file correctly or some other debug that might shed light on why it is not working for me ?

**Reply by Lailloken** [reply]
OMG, the thousands separators are not tied to the language-setting of the game but the language/region setting of the OS. It's using periods on my end and commas on yours. That's so random, and it explains why I couldn't find the issue in the code.
EDIT: updated the dev build, should be fixed

**Reply by GrimDawn-max** [reply]
> OMG, the thousands separators are not tied to the language-setting of the game but the language/region setting of the OS. It's using periods on my end and commas on yours. That's so random, and it explains why I couldn't find the issue in the code. EDIT: updated the dev build, should be fixed

Well spotted. That's a nasty one. I downloaded the update and tested it. It's all working now. I get the kill count displayed with the loot list and the same number is now included in the export file. Thanks again for putting this together !

**Comment by GrimDawn-max** [comment]
OK. I’ll grab it in a bit and try it out
